File: code/lt0141.java
// https://leetcode-cn.com/problems/linked-list-cycle/
// linked list cycle
/**
 * Definition for singly-linked list.
 * class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public boolean hasCycle(ListNode head) {
        if(head == null)
            return false;
        ListNode slow = head;
        if(slow.next == null)
            return false;
        ListNode fast = slow.next;
        while(fast != slow){
            if(fast.next == null || fast.next.next == null)
                return false;
            fast = fast.next.next;
            slow = slow.next;
        }
        return true;
    }
}

File: code/lt0141.cpp
struct ListNode
{
    int val;
    ListNode* next;
    ListNode(int x):val(x), next(nullptr){};
};

class Solution {
public:
    bool hasCycle(ListNode *head) {
        if(head == nullptr)
            return false;
        
        if(head->next == nullptr)
            return false;

        ListNode* slow = head, *fast = head->next;

        while(slow && fast && slow != fast)
        {
            slow = slow->next;
            fast = fast->next;
            if(fast)
                fast = fast->next;
        }

        if(slow == nullptr || fast == nullptr)
            return false;
        return true;
    }
};